File: python/basic/object_detection.py
"""
目标检测相关问题：
71. 滑动窗口检测 - 使用滑动窗口进行目标检测
72. HOG+SVM检测 - 使用HOG特征和SVM进行目标检测
73. Haar+AdaBoost检测 - 使用Haar特征和AdaBoost进行目标检测
74. 非极大值抑制 - 实现非极大值抑制算法
75. 目标跟踪 - 实现简单的目标跟踪算法
"""

import cv2
import numpy as np
from sklearn import svm
from sklearn.preprocessing import StandardScaler

def sliding_window_detection(img_path, window_size=(64, 64), stride=32):
    """
    问题71：滑动窗口检测
    使用滑动窗口进行目标检测

    参数:
        img_path: 输入图像路径
        window_size: 窗口大小，默认(64, 64)
        stride: 步长，默认32

    返回:
        检测结果可视化
    """
    # 读取图像
    img = cv2.imread(img_path)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 复制原图用于绘制结果
    result = img.copy()

    # 转换为灰度图
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    # 定义简单的目标检测函数（这里使用边缘检测作为示例）
    def detect_object(window):
        edges = cv2.Canny(window, 100, 200)
        return np.sum(edges) > 1000  # 简单的阈值判断

    # 滑动窗口检测
    for y in range(0, img.shape[0] - window_size[1], stride):
        for x in range(0, img.shape[1] - window_size[0], stride):
            # 提取窗口
            window = gray[y:y+window_size[1], x:x+window_size[0]]

            # 检测目标
            if detect_object(window):
                # 绘制检测框
                cv2.rectangle(result, (x, y),
                            (x + window_size[0], y + window_size[1]),
                            (0, 255, 0), 2)

    return result

class HOGExtractor:
    """HOG特征提取器"""
    def __init__(self, win_size=(64, 128), block_size=(16, 16),
                 block_stride=(8, 8), cell_size=(8, 8), nbins=9):
        self.win_size = win_size
        self.block_size = block_size
        self.block_stride = block_stride
        self.cell_size = cell_size
        self.nbins = nbins

        # 计算特征维度
        n_cells = (win_size[0] // cell_size[0], win_size[1] // cell_size[1])
        n_blocks = ((n_cells[0] - block_size[0] // cell_size[0]) // (block_stride[0] // cell_size[0]) + 1,
                   (n_cells[1] - block_size[1] // cell_size[1]) // (block_stride[1] // cell_size[1]) + 1)
        self.feature_dim = n_blocks[0] * n_blocks[1] * block_size[0] * block_size[1] * nbins // (cell_size[0] * cell_size[1])

    def compute_gradients(self, img):
        """计算图像梯度"""
        # 确保图像是灰度图
        if len(img.shape) > 2:
            img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

        # 计算x和y方向的梯度
        gx = cv2.Sobel(img, cv2.CV_32F, 1, 0, ksize=1)
        gy = cv2.Sobel(img, cv2.CV_32F, 0, 1, ksize=1)

        # 计算梯度幅值和方向
        magnitude = np.sqrt(gx**2 + gy**2)
        angle = np.arctan2(gy, gx) * 180 / np.pi
        angle[angle < 0] += 180

        return magnitude, angle

    def compute_cell_histogram(self, magnitude, angle):
        """计算cell直方图"""
        n_cells = (self.win_size[0] // self.cell_size[0],
                  self.win_size[1] // self.cell_size[1])
        cell_hists = np.zeros((n_cells[1], n_cells[0], self.nbins))

        # 计算每个像素的贡献
        for y in range(magnitude.shape[0]):
            for x in range(magnitude.shape[1]):
                mag = magnitude[y, x]
                ang = angle[y, x]

                # 计算bin索引
                bin_width = 180.0 / self.nbins
                bin_idx = int(ang / bin_width)
                next_bin = (bin_idx + 1) % self.nbins
                alpha = (ang - bin_idx * bin_width) / bin_width

                # 计算cell索引
                cell_x = x // self.cell_size[0]
                cell_y = y // self.cell_size[1]

                if cell_x < n_cells[0] and cell_y < n_cells[1]:
                    cell_hists[cell_y, cell_x, bin_idx] += mag * (1 - alpha)
                    cell_hists[cell_y, cell_x, next_bin] += mag * alpha

        return cell_hists

    def compute_block_features(self, cell_hists):
        """计算block特征并归一化"""
        n_cells = (self.win_size[0] // self.cell_size[0],
                  self.win_size[1] // self.cell_size[1])
        n_blocks = ((n_cells[0] - self.block_size[0] // self.cell_size[0]) //
                   (self.block_stride[0] // self.cell_size[0]) + 1,
                   (n_cells[1] - self.block_size[1] // self.cell_size[1]) //
                   (self.block_stride[1] // self.cell_size[1]) + 1)

        features = []
        for by in range(0, n_cells[1] - self.block_size[1] // self.cell_size[1] + 1,
                       self.block_stride[1] // self.cell_size[1]):
            for bx in range(0, n_cells[0] - self.block_size[0] // self.cell_size[0] + 1,
                          self.block_stride[0] // self.cell_size[0]):
                # 提取block中的cell直方图
                block_features = cell_hists[by:by + self.block_size[1] // self.cell_size[1],
                                         bx:bx + self.block_size[0] // self.cell_size[0]].flatten()

                # L2-Norm归一化
                norm = np.sqrt(np.sum(block_features**2) + 1e-5)
                block_features = block_features / norm

                features.extend(block_features)

        return np.array(features)

    def compute(self, img):
        """计算HOG特征"""
        # 调整图像大小
        img = cv2.resize(img, self.win_size)

        # 计算梯度
        magnitude, angle = self.compute_gradients(img)

        # 计算cell直方图
        cell_hists = self.compute_cell_histogram(magnitude, angle)

        # 计算block特征并归一化
        features = self.compute_block_features(cell_hists)

        return features

def hog_svm_detection(img_path, window_size=(64, 64), stride=32):
    """
    问题72：HOG+SVM检测
    使用HOG特征和SVM进行目标检测

    参数:
        img_path: 输入图像路径
        window_size: 窗口大小，默认(64, 64)
        stride: 步长，默认32

    返回:
        检测结果可视化
    """
    # 读取图像
    img = cv2.imread(img_path)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 复制原图用于绘制结果
    result = img.copy()
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    # 创建HOG特征提取器
    hog = HOGExtractor(win_size=window_size)

    # 创建和训练SVM（这里使用简单的示例数据）
    svm_classifier = svm.LinearSVC()
    scaler = StandardScaler()

    # 生成一些示例数据进行训练
    n_samples = 100
    n_features = hog.feature_dim
    X = np.random.randn(n_samples, n_features)
    y = np.random.randint(0, 2, n_samples)

    # 训练SVM
    X_scaled = scaler.fit_transform(X)
    svm_classifier.fit(X_scaled, y)

    # 滑动窗口检测
    for y in range(0, img.shape[0] - window_size[1], stride):
        for x in range(0, img.shape[1] - window_size[0], stride):
            # 提取窗口
            window = gray[y:y+window_size[1], x:x+window_size[0]]

            # 计算HOG特征
            features = hog.compute(window)

            # 特征标准化
            features_scaled = scaler.transform(features.reshape(1, -1))

            # SVM预测
            if svm_classifier.predict(features_scaled)[0] == 1:
                # 绘制检测框
                cv2.rectangle(result, (x, y),
                            (x + window_size[0], y + window_size[1]),
                            (0, 255, 0), 2)

    return result

def haar_adaboost_detection(img_path):
    """
    问题73：Haar+AdaBoost检测
    使用Haar特征和AdaBoost进行目标检测

    参数:
        img_path: 输入图像路径

    返回:
        检测结果可视化
    """
    # 读取图像
    img = cv2.imread(img_path)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 复制原图用于绘制结果
    result = img.copy()

    # 转换为灰度图
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    # 加载预训练的人脸检测器（作为示例）
    face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_frontalface_default.xml')

    # 进行人脸检测
    faces = face_cascade.detectMultiScale(gray, 1.1, 4)

    # 绘制检测结果
    for (x, y, w, h) in faces:
        cv2.rectangle(result, (x, y), (x+w, y+h), (0, 255, 0), 2)

    return result

def compute_nms_manual(boxes, scores, iou_threshold=0.5):
    """手动实现非极大值抑制算法

    参数:
        boxes: 边界框坐标列表，每个框为[x1, y1, x2, y2]格式
        scores: 每个边界框对应的置信度分数
        iou_threshold: IoU阈值，默认0.5

    返回:
        keep: 保留的边界框索引列表
    """
    # 确保输入数据格式正确
    boxes = np.array(boxes)
    scores = np.array(scores)

    # 计算所有框的面积
    x1 = boxes[:, 0]
    y1 = boxes[:, 1]
    x2 = boxes[:, 2]
    y2 = boxes[:, 3]
    areas = (x2 - x1) * (y2 - y1)

    # 根据分数排序
    order = scores.argsort()[::-1]

    keep = []
    while order.size > 0:
        # 保留分数最高的框
        i = order[0]
        keep.append(i)

        if order.size == 1:
            break

        # 计算其他框与当前框的IoU
        xx1 = np.maximum(x1[i], x1[order[1:]])
        yy1 = np.maximum(y1[i], y1[order[1:]])
        xx2 = np.minimum(x2[i], x2[order[1:]])
        yy2 = np.minimum(y2[i], y2[order[1:]])

        w = np.maximum(0.0, xx2 - xx1)
        h = np.maximum(0.0, yy2 - yy1)
        inter = w * h

        # 计算IoU
        ovr = inter / (areas[i] + areas[order[1:]] - inter)

        # 保留IoU小于阈值的框
        inds = np.where(ovr <= iou_threshold)[0]
        order = order[inds + 1]

    return keep

def object_tracking(img_path, video_path=None):
    """
    问题75：目标跟踪
    实现简单的目标跟踪算法

    参数:
        img_path: 输入图像路径
        video_path: 视频路径（可选）

    返回:
        跟踪结果可视化
    """
    # 读取图像
    img = cv2.imread(img_path)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 复制原图用于绘制结果
    result = img.copy()

    # 转换为HSV颜色空间
    hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)

    # 定义目标颜色范围（以红色为例）
    lower_red = np.array([0, 50, 50])
    upper_red = np.array([10, 255, 255])

    # 创建掩码
    mask = cv2.inRange(hsv, lower_red, upper_red)

    # 寻找轮廓
    contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    if contours:
        # 找到最大的轮廓
        c = max(contours, key=cv2.contourArea)

        # 计算边界框
        x, y, w, h = cv2.boundingRect(c)

        # 绘制跟踪框
        cv2.rectangle(result, (x, y), (x+w, y+h), (0, 255, 0), 2)

        # 绘制目标中心
        center = (x + w//2, y + h//2)
        cv2.circle(result, center, 5, (0, 0, 255), -1)

    return result

def main(problem_id=71):
    """
    主函数，通过problem_id选择要运行的问题

    参数:
        problem_id: 问题编号，默认为71
    """
    input_path = "../images/imori.jpg"

    if problem_id == 71:
        result = sliding_window_detection(input_path)
        output_path = "../images/answer_71.jpg"
        title = "滑动窗口检测"
    elif problem_id == 72:
        result = hog_svm_detection(input_path)
        output_path = "../images/answer_72.jpg"
        title = "HOG+SVM检测"
    elif problem_id == 73:
        result = haar_adaboost_detection(input_path)
        output_path = "../images/answer_73.jpg"
        title = "Haar+AdaBoost检测"
    elif problem_id == 74:
        # 示例边界框和分数
        boxes = [[100, 100, 200, 200], [150, 150, 250, 250], [120, 120, 220, 220]]
        scores = [0.9, 0.8, 0.7]
        keep = compute_nms_manual(boxes, scores)

        # 在原图上绘制保留的边界框
        result = cv2.imread(input_path)
        for i in keep:
            x1, y1, x2, y2 = boxes[i]
            cv2.rectangle(result, (x1, y1), (x2, y2), (0, 255, 0), 2)

        output_path = "../images/answer_74.jpg"
        title = "非极大值抑制"
    elif problem_id == 75:
        result = object_tracking(input_path)
        output_path = "../images/answer_75.jpg"
        title = "目标跟踪"
    else:
        print(f"问题 {problem_id} 不存在")
        return

    # 保存结果
    cv2.imwrite(output_path, result)

    # 显示结果
    cv2.imshow(title, result)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

if __name__ == "__main__":
    import sys
    problem_id = int(sys.argv[1]) if len(sys.argv) > 1 else 71
    main(problem_id)

File: cpp/basic/object_detection.cpp
#include <basic/object_detection.hpp>
#include <opencv2/opencv.hpp>
#include <opencv2/ml.hpp>  // 添加 ml 模块支持 SVM
#include <omp.h>
#include <numeric>  // 添加 numeric 支持 iota

namespace ip101 {

using namespace cv;
using namespace std;

namespace {
// Internal constants
constexpr int CACHE_LINE = 64;    // CPU cache line size (bytes)
constexpr int BLOCK_SIZE = 16;    // Block processing size

// HOG feature extractor
class HOGExtractor {
public:
    HOGExtractor(Size win_size = Size(64, 128),
                Size block_size = Size(16, 16),
                Size block_stride = Size(8, 8),
                Size cell_size = Size(8, 8),
                int nbins = 9)
        : win_size_(win_size),
          block_size_(block_size),
          block_stride_(block_stride),
          cell_size_(cell_size),
          nbins_(nbins) {
        // Calculate HOG feature dimension
        Size n_cells(win_size.width / cell_size.width,
                    win_size.height / cell_size.height);
        Size n_blocks((n_cells.width - block_size.width / cell_size.width) /
                     (block_stride.width / cell_size.width) + 1,
                     (n_cells.height - block_size.height / cell_size.height) /
                     (block_stride.height / cell_size.height) + 1);
        feature_dim_ = n_blocks.width * n_blocks.height *
                      block_size.width * block_size.height *
                      nbins / (cell_size.width * cell_size.height);
    }

    // Calculate image gradients
    void computeGradients(const Mat& img, Mat& magnitude, Mat& angle) const {
        magnitude = Mat::zeros(img.size(), CV_32F);
        angle = Mat::zeros(img.size(), CV_32F);

        #pragma omp parallel for
        for (int y = 1; y < img.rows - 1; y++) {
            for (int x = 1; x < img.cols - 1; x++) {
                // Calculate x-direction gradient
                float gx = static_cast<float>(img.at<uchar>(y, x + 1) - img.at<uchar>(y, x - 1));
                // Calculate y-direction gradient
                float gy = static_cast<float>(img.at<uchar>(y + 1, x) - img.at<uchar>(y - 1, x));

                // Calculate gradient magnitude
                magnitude.at<float>(y, x) = std::sqrt(gx * gx + gy * gy);
                // Calculate gradient direction (angle)
                angle.at<float>(y, x) = static_cast<float>(std::atan2(gy, gx) * 180.0 / CV_PI);
                if (angle.at<float>(y, x) < 0) {
                    angle.at<float>(y, x) += 180.0f;
                }
            }
        }
    }

    // Calculate cell histogram
    void computeCellHistogram(const Mat& magnitude, const Mat& angle,
                            vector<vector<vector<float>>>& cell_hists) const {
        Size n_cells(win_size_.width / cell_size_.width,
                    win_size_.height / cell_size_.height);

        #pragma omp parallel for
        for (int y = 0; y < magnitude.rows; y++) {
            for (int x = 0; x < magnitude.cols; x++) {
                float mag = magnitude.at<float>(y, x);
                float ang = angle.at<float>(y, x);

                // Calculate bin index
                float bin_width = 180.0f / nbins_;
                int bin = static_cast<int>(ang / bin_width);
                int next_bin = (bin + 1) % nbins_;
                float alpha = (ang - bin * bin_width) / bin_width;

                // Calculate cell index
                int cell_x = x / cell_size_.width;
                int cell_y = y / cell_size_.height;

                if (cell_x < n_cells.width && cell_y < n_cells.height) {
                    #pragma omp atomic
                    cell_hists[cell_y][cell_x][bin] += mag * (1 - alpha);
                    #pragma omp atomic
                    cell_hists[cell_y][cell_x][next_bin] += mag * alpha;
                }
            }
        }
    }

    // Calculate block features and normalize
    void computeBlockFeatures(const vector<vector<vector<float>>>& cell_hists,
                            vector<float>& features) const {
        Size n_cells(win_size_.width / cell_size_.width,
                    win_size_.height / cell_size_.height);
        Size n_blocks((n_cells.width - block_size_.width / cell_size_.width) /
                     (block_stride_.width / cell_size_.width) + 1,
                     (n_cells.height - block_size_.height / cell_size_.height) /
                     (block_stride_.height / cell_size_.height) + 1);

        for (int by = 0; by <= n_cells.height - block_size_.height / cell_size_.height;
             by += block_stride_.height / cell_size_.height) {
            for (int bx = 0; bx <= n_cells.width - block_size_.width / cell_size_.width;
                 bx += block_stride_.width / cell_size_.width) {
                vector<float> block_features;
                float norm = 0;

                // Collect all cell histograms in the block
                for (int cy = by; cy < by + block_size_.height / cell_size_.height; cy++) {
                    for (int cx = bx; cx < bx + block_size_.width / cell_size_.width; cx++) {
                        block_features.insert(block_features.end(),
                                           cell_hists[cy][cx].begin(),
                                           cell_hists[cy][cx].end());
                        for (float val : cell_hists[cy][cx]) {
                            norm += val * val;
                        }
                    }
                }

                // L2-Norm normalization
                norm = std::sqrt(norm + 1e-5f);
                for (float& val : block_features) {
                    val /= norm;
                }

                features.insert(features.end(),
                              block_features.begin(),
                              block_features.end());
            }
        }
    }

    vector<float> compute(const Mat& img) const {
        // Resize image
        Mat resized;
        resize(img, resized, win_size_);

        // Calculate gradients
        Mat magnitude, angle;
        computeGradients(resized, magnitude, angle);

        // Calculate cell histograms
        Size n_cells(win_size_.width / cell_size_.width,
                    win_size_.height / cell_size_.height);
        vector<vector<vector<float>>> cell_hists(n_cells.height,
            vector<vector<float>>(n_cells.width, vector<float>(nbins_, 0)));
        computeCellHistogram(magnitude, angle, cell_hists);

        // Calculate block features and normalize
        vector<float> features;
        features.reserve(feature_dim_);
        computeBlockFeatures(cell_hists, features);

        return features;
    }

private:
    Size win_size_;
    Size block_size_;
    Size block_stride_;
    Size cell_size_;
    int nbins_;
    int feature_dim_;
};

// Haar feature extractor
class HaarExtractor {
public:
    HaarExtractor() {
        // Load pre-trained Haar cascade classifier
        face_cascade_.load("haarcascade_frontalface_alt.xml");
    }

    vector<Rect> detect(const Mat& img, double scale_factor = 1.1, int min_neighbors = 3) {
        vector<Rect> faces;
        face_cascade_.detectMultiScale(img, faces, scale_factor, min_neighbors);
        return faces;
    }

private:
    CascadeClassifier face_cascade_;
};

// Calculate IoU between two rectangles
float compute_iou(const Rect& a, const Rect& b) {
    int x1 = max(a.x, b.x);
    int y1 = max(a.y, b.y);
    int x2 = min(a.x + a.width, b.x + b.width);
    int y2 = min(a.y + a.height, b.y + b.height);

    if (x1 >= x2 || y1 >= y2) return 0.0f;

    float intersection_area = static_cast<float>((x2 - x1) * (y2 - y1));
    float union_area = static_cast<float>(a.width * a.height + b.width * b.height - intersection_area);

    return intersection_area / union_area;
}

} // anonymous namespace

vector<DetectionResult> sliding_window_detect(
    const Mat& src,
    const Size& window_size,
    int stride,
    float threshold) {

    vector<DetectionResult> results;
    HOGExtractor hog(window_size);

    // Load pre-trained SVM model
    Ptr<ml::SVM> svm = ml::SVM::load("pedestrian_svm.xml");

    #pragma omp parallel for
    for (int y = 0; y <= src.rows - window_size.height; y += stride) {
        for (int x = 0; x <= src.cols - window_size.width; x += stride) {
            // Extract window
            Mat window = src(Rect(x, y, window_size.width, window_size.height));

            // Calculate HOG features
            vector<float> features = hog.compute(window);

            // SVM prediction
            Mat feature_mat(1, static_cast<int>(features.size()), CV_32F);
            memcpy(feature_mat.data, features.data(), features.size() * sizeof(float));
            float score = svm->predict(feature_mat, noArray(), ml::StatModel::RAW_OUTPUT);

            if (score > threshold) {
                DetectionResult det;
                det.bbox = Rect(x, y, window_size.width, window_size.height);
                det.confidence = score;
                det.class_id = 1;  // pedestrian class
                det.label = "pedestrian";

                #pragma omp critical
                results.push_back(det);
            }
        }
    }

    return results;
}

vector<DetectionResult> hog_svm_detect(const Mat& src, float threshold) {
    vector<DetectionResult> results;

    // Use OpenCV built-in HOG detector
    HOGDescriptor hog;
    hog.setSVMDetector(HOGDescriptor::getDefaultPeopleDetector());

    // Detect pedestrians
    vector<Rect> found;
    vector<double> weights;
    hog.detectMultiScale(src, found, weights,
                        0,              // hit threshold
                        Size(8,8),      // win stride
                        Size(32,32),    // padding
                        1.05,           // scale
                        threshold);     // final threshold

    // Convert result format
    for (size_t i = 0; i < found.size(); i++) {
        DetectionResult det;
        det.bbox = found[i];
        det.confidence = static_cast<float>(weights[i]);
        det.class_id = 1;  // pedestrian class
        det.label = "pedestrian";
        results.push_back(det);
    }

    return results;
}

vector<DetectionResult> haar_face_detect(const Mat& src, float threshold) {
    vector<DetectionResult> results;

    // Create Haar feature extractor
    HaarExtractor haar;

    // Detect faces
    vector<Rect> faces = haar.detect(src, 1.1, 3);

    // Convert result format
    for (const auto& face : faces) {
        DetectionResult det;
        det.bbox = face;
        det.confidence = 1.0f;  // Haar classifier has no confidence output
        det.class_id = 2;  // face class
        det.label = "face";
        results.push_back(det);
    }

    return results;
}

vector<int> nms(const vector<Rect>& boxes,
                const vector<float>& scores,
                float iou_threshold) {

    vector<int> indices(boxes.size());
    std::iota(indices.begin(), indices.end(), 0);  // Fill with 0, 1, 2, ...

    // Sort by confidence in descending order
    sort(indices.begin(), indices.end(),
         [&scores](int a, int b) { return scores[a] > scores[b]; });

    vector<int> keep;
    while (!indices.empty()) {
        int curr = indices[0];
        keep.push_back(curr);

        vector<int> tmp;
        for (size_t i = 1; i < indices.size(); i++) {
            float iou = compute_iou(boxes[curr], boxes[indices[i]]);
            if (iou <= iou_threshold) {
                tmp.push_back(indices[i]);
            }
        }
        indices = tmp;
    }

    return keep;
}

vector<DetectionResult> track_objects(
    const Mat& src,
    const Mat& prev,
    const vector<DetectionResult>& prev_boxes) {

    vector<DetectionResult> curr_boxes;

    // Calculate optical flow
    vector<Point2f> prev_points, curr_points;
    for (const auto& det : prev_boxes) {
        prev_points.push_back(Point2f(static_cast<float>(det.bbox.x + det.bbox.width/2),
                                    static_cast<float>(det.bbox.y + det.bbox.height/2)));
    }

    vector<uchar> status;
    vector<float> err;
    calcOpticalFlowPyrLK(prev, src, prev_points, curr_points, status, err);

    // Update detection box positions
    for (size_t i = 0; i < prev_boxes.size(); i++) {
        if (status[i]) {
            DetectionResult det = prev_boxes[i];
            float dx = curr_points[i].x - prev_points[i].x;
            float dy = curr_points[i].y - prev_points[i].y;
            det.bbox.x += static_cast<int>(dx);
            det.bbox.y += static_cast<int>(dy);
            curr_boxes.push_back(det);
        }
    }

    return curr_boxes;
}

Mat draw_detections(const Mat& src,
                   const vector<DetectionResult>& detections) {

    Mat img_display = src.clone();
    if (img_display.channels() == 1) {
        cvtColor(img_display, img_display, COLOR_GRAY2BGR);
    }

    for (const auto& det : detections) {
        Scalar color;
        if (det.class_id == 1) {  // pedestrian
            color = Scalar(0, 255, 0);
        } else if (det.class_id == 2) {  // face
            color = Scalar(0, 0, 255);
        } else {
            color = Scalar(255, 0, 0);
        }

        // Draw bounding box
        rectangle(img_display, det.bbox, color, 2);

        // Draw label and confidence
        string label = format("%s %.2f", det.label.c_str(), det.confidence);
        int baseline = 0;
        Size text_size = getTextSize(label, FONT_HERSHEY_SIMPLEX,
                                   0.5, 1, &baseline);
        rectangle(img_display,
                 Point(det.bbox.x, det.bbox.y - text_size.height - 5),
                 Point(det.bbox.x + text_size.width, det.bbox.y),
                 color, -1);
        putText(img_display, label,
                Point(det.bbox.x, det.bbox.y - 5),
                FONT_HERSHEY_SIMPLEX, 0.5, Scalar(255, 255, 255), 1);
    }

    return img_display;
}

} // namespace ip101